test/e2e/testenv/testenv.go

```diff
@@ -0,0 +1,659 @@
+	}
+
+	return &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", jobNameEtcdLoader, testutils.GenerateRandomAlphanumericString(g, 4)),
+			Namespace: namespace,
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "loader",
+							Image:           "alpine/curl",
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command:         []string{"/bin/sh", "-c"},
+							Args:            []string{getEtcdLoaderScript(numKeys, etcdClientServiceName, etcdClientServicePort)},
+							VolumeMounts: []corev1.VolumeMount{
+								{MountPath: "/var/etcd/ssl/ca", Name: "client-url-ca-etcd"},
+								{MountPath: "/var/etcd/ssl/client", Name: "client-url-etcd-client-tls"},
+							},
+						},
+					},
+					RestartPolicy: corev1.RestartPolicyNever,
+					Volumes: []corev1.Volume{
+						getTLSVolume("client-url-ca-etcd", etcdClientTLS.TLSCASecretRef.Name),
+						getTLSVolume("client-url-etcd-client-tls", etcdClientTLS.ClientTLSSecretRef.Name),
+					},
+				},
+			},
+			BackoffLimit: ptr.To[int32](0),
+		},
+	}, nil
+}
+
+// getEtcdLoaderScript generates the shell script used by the etcd loader job to populate etcd with keys.
+func getEtcdLoaderScript(numKeys int64, etcdClientServiceName string, etcdClientServicePort int32) string {
+	return fmt.Sprintf(`
+	for i in $(seq 1 %d); do
+		key=$(echo -n "key-$i" | base64)
+		value=$(echo -n "value-$i" | base64)
+		curl -s -X POST \
+		-H "Content-Type: application/json" \
+		--cacert /var/etcd/ssl/ca/ca.crt \
+		--cert /var/etcd/ssl/client/tls.crt \
+		--key /var/etcd/ssl/client/tls.key \
+		-d "{\"key\": \"$key\", \"value\": \"$value\"}" \
+		-L https://%s:%d/v3/kv/put;
+	done
+	`, numKeys, etcdClientServiceName, etcdClientServicePort)
+}
+
+// getTLSVolume creates a volume for the TLS secret.
+func getTLSVolume(name, secretName string) corev1.Volume {
+	return corev1.Volume{
+		Name: name,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName:  secretName,
+				DefaultMode: ptr.To(common.ModeOwnerReadWriteGroupRead),
+			},
+		},
+	}
+}
+
+// TriggerFullSnapshot triggers a full snapshot of the etcd cluster using etcd-backup-restore server's HTTP API.
+func (t *TestEnvironment) TriggerFullSnapshot(g *WithT, etcd *druidv1alpha1.Etcd, timeout time.Duration) {
+	t.TriggerSnapshot(g, etcd, "full", timeout)
+}
+
+// TriggerDeltaSnapshot triggers a delta snapshot of the etcd cluster using etcd-backup-restore server's HTTP API.
+func (t *TestEnvironment) TriggerDeltaSnapshot(g *WithT, etcd *druidv1alpha1.Etcd, timeout time.Duration) {
+	t.TriggerSnapshot(g, etcd, "delta", timeout)
+}
+
+// TriggerSnapshot triggers a snapshot of the etcd cluster using etcd-backup-restore server's HTTP API.
+func (t *TestEnvironment) TriggerSnapshot(g *WithT, etcd *druidv1alpha1.Etcd, snapshotType string, timeout time.Duration) {
```

Comment:
Just a thought: the notion of "triggering" a snapshot in both `etcd-druid` and `etcd-backup-restore` codebases suggests that the process of taking a snapshot has just been started, when in reality the API call to `/snapshot/*` in `backup-restore` only returns after a snapshot has been taken (the process has completed at this point).